File: hse/BPM161/alferov/2/src/main/java/net/vasalf/hw/lockfree/LockFreeSet.java
package net.vasalf.hw.lockfree;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.concurrent.atomic.AtomicMarkableReference;

public class LockFreeSet<T extends Comparable<T>> implements MySet<T> {
    private static class Node<T extends Comparable<T>> {
        private T value;
        private AtomicMarkableReference<Node<T>> next;

        Node(T value, Node<T> next) {
            this.value = value;
            this.next = new AtomicMarkableReference<>(next, false);
        }

        public T getValue() {
            return value;
        }

        public AtomicMarkableReference<Node<T>> getNext() {
            return next;
        }
    }

    private Node<T> head;
    private Node<T> tail;

    public LockFreeSet() {
        tail = new Node<>(null, null);
        head = new Node<>(null, tail);
    }

    @Override
    public boolean add(T value) {
        while (true) {
            Node<T> prev = null;
            boolean prevMarked = false;
            Node<T> cur = head;
            boolean[] curMarked = new boolean[1];
            Node<T> curNext = head.getNext().get(curMarked);
            while (cur != tail) {
                if (cur != head && cur.getValue().compareTo(value) == 0 && !curMarked[0]) {
                    return false;
                }
                prev = cur;
                prevMarked = curMarked[0];
                cur = curNext;
                curNext = cur.getNext().get(curMarked);
            }
            Node<T> newNode = new Node<>(value, tail);
            if (prev.getNext().compareAndSet(tail, newNode, prevMarked, prevMarked)) {
                return true;
            }
        }
    }

    @Override
    public boolean remove(T value) {
        while (true) {
            boolean rep = false;
            Node<T> cur = head;
            boolean[] curMarked = new boolean[1];
            Node<T> curNext = head.getNext().get(curMarked);
            while (cur != tail) {
                if (cur != head && cur.getValue().compareTo(value) == 0 && !curMarked[0]) {
                    if (cur.getNext().compareAndSet(curNext, curNext, false, true)) {
                        return true;
                    } else {
                        rep = true;
                        break;
                    }
                }
                cur = curNext;
                curNext = cur.getNext().get(curMarked);
            }
            if (!rep) {
                return false;
            }
        }
    }

    @Override
    public boolean contains(T value) {
        Node<T> cur = head;
        boolean[] curMarked = new boolean[1];
        Node<T> curNext = head.getNext().get(curMarked);
        while (cur != tail) {
            if (cur != head && cur.getValue().compareTo(value) == 0 && !curMarked[0]) {
                return true;
            }
            cur = curNext;
            curNext = cur.getNext().get(curMarked);
        }
        return false;
    }

    @Override
    public boolean isEmpty() {
        Node<T> cur = head;
        boolean[] curMarked = new boolean[1];
        Node<T> curNext = head.getNext().get(curMarked);
        while (cur != tail) {
            if (cur != head && !curMarked[0]) {
                return false;
            }
            cur = curNext;
            curNext = cur.getNext().get(curMarked);
        }
        return true;
    }

    @Override
    public Iterator<T> iterator() {
        List<T> prev = null;
        List<T> cur = new ArrayList<>();

        do {
            prev = new ArrayList<>(cur);
            cur.clear();
            Node<T> node = head;
            boolean[] marked = new boolean[1];
            while (node != tail) {
                Node<T> next = node.getNext().get(marked);
                if (node != head && !marked[0]) {
                    cur.add(node.value);
                }
                node = next;
            }
        } while (!prev.equals(cur));

        return cur.iterator();
    }
}

File: hse/BPM161/vasilenko/2/src/main/kotlin/lockfree/LockFreeSet.kt
package lockfree

import java.util.concurrent.atomic.AtomicMarkableReference

typealias LockFreeNode<T> = AtomicMarkableReference<LockFreeSet.Node<T>>

class LockFreeSet<T> : ILockFreeSet<T>
        where T : Comparable<T> {

    class Node<T>(val value: T?, var next: LockFreeNode<T>)

    private val tail = LockFreeNode<T>(null, false)
    private var head = LockFreeNode(Node(null, tail), false)

    override val isEmpty: Boolean
        get() {
            val mark = booleanArrayOf(false)
            tryDeleteAfter(head)
            var cur = head.reference.next
            var curNode = cur.get(mark)
            while (curNode != null) {
                tryDeleteAfter(cur)
                if (!mark[0]) {
                    return false
                }
                cur = curNode.next
                curNode = cur.get(mark)
            }
            return true
        }

    override fun add(value: T): Boolean {
        while (true) {
            var cur = head
            val curMark = booleanArrayOf(false)
            var curNode = cur.reference
            tryDeleteAfter(cur)
            val nextMark = booleanArrayOf(false)
            var nextNode = curNode.next.get(nextMark)
            while (nextNode != null) {
                if (nextNode.value == value && !nextMark[0]) {
                    return false
                }
                cur = curNode.next
                curNode = nextNode
                curMark[0] = nextMark[0]
                tryDeleteAfter(cur)
                nextNode = curNode.next.get(nextMark)
            }
            if (cur.compareAndSet(
                    curNode,
                    Node(curNode.value, LockFreeNode(Node(value, tail), false)),
                    curMark[0],
                    curMark[0]
                )
            ) {
                return true
            }
        }
    }

    override fun remove(value: T): Boolean {
        val mark = booleanArrayOf(false)
        tryDeleteAfter(head)
        var cur = head.reference.next
        var curNode = cur.get(mark)
        while (curNode != null) {
            while (curNode.value == value && !mark[0]) {
                if (cur.compareAndSet(curNode, curNode, false, true)) {
                    return true
                }
                curNode = cur.get(mark)
            }
            tryDeleteAfter(cur)
            cur = curNode.next
            curNode = cur.get(mark)
        }
        return false
    }

    override operator fun contains(value: T): Boolean {
        val mark = booleanArrayOf(false)
        tryDeleteAfter(head)
        var cur = head.reference.next
        var curNode = cur.get(mark)
        while (curNode != null) {
            tryDeleteAfter(cur)
            if (curNode.value == value && !mark[0]) {
                return true
            }
            cur = curNode.next
            curNode = cur.get(mark)
        }
        return false
    }

    override operator fun iterator(): Iterator<T> {
        var curSeq = mutableListOf<Node<T>>()
        var prevSeq = mutableListOf<Node<T>>()
        do {
            curSeq.clear()
            val mark = booleanArrayOf(false)
            tryDeleteAfter(head)
            var cur = head.reference.next
            var curNode = cur.get(mark)
            while (curNode != null) {
                tryDeleteAfter(cur)
                if (!mark[0]) {
                    curSeq.add(curNode)
                }
                cur = curNode.next
                curNode = cur.get(mark)
            }
            curSeq = prevSeq.also { prevSeq = curSeq }
        } while (curSeq != prevSeq)
        return curSeq.map { it.value!! }.iterator()
    }

    private fun tryDeleteAfter(cur: LockFreeNode<T>) {
        val mark = booleanArrayOf(false)
        val curNode = cur.get(mark)
        val nextMark = booleanArrayOf(false)
        val nextNode = curNode.next.get(nextMark)
        val nextNextMark = booleanArrayOf(false)

        // Current node was not removed, next nde marked to remove, next node is not last
        if (!mark[0] && nextMark[0] && nextNode.next.get(nextNextMark) != null) {
            cur.compareAndSet(curNode, Node(curNode.value, nextNode.next), false, false)
        }
    }
}

